### PR TITLE
"script-to-upload-evidence-to-cloud-create-spotlightData"

### PR DESCRIPTION
**Pull request by MallanagoudaB**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

File: Project-Service-implementation-Script/common_config/config.ini [hunk 1]
```diff
@@ -312,6 +318,9 @@ readcourseurl = api/content/v1/read/
 # Endpoint for fetching organization details
 fetchorgdetails = api/org/v1/search
 
+# Endpoint for fetching URL
+fetchPreSignedUrl = project/v1/cloud-services/files/preSignedUrls
```
Comment (adithyadinesh0412) [inline Project-Service-implementation-Script/common_config/config.ini:322]: Make it standard. @MallanagoudaB

File: Project-Service-implementation-Script/main.py [hunk 2]
```diff
@@ -38,6 +38,8 @@
 from openpyxl.styles import colors
 from openpyxl.cell import Cell
 import gdown
+import requests
```
Comment (adithyadinesh0412) [inline Project-Service-implementation-Script/main.py:41]: Duplicate import. Please remove @MallanagoudaB

File: Project-Service-implementation-Script/main.py [hunk 3]
```diff
@@ -3430,9 +3431,240 @@ def prepareProjectAndTasksSheets(project_inputFile, projectName_for_folder_path,
         with open(taskFilePath + 'taskUpload.csv','a',encoding='utf-8') as file:
             writer = csv.writer(file, quoting=csv.QUOTE_NONNUMERIC, delimiter=',',lineterminator='\n')
             writer.writerows([task_values])
-        #   subtaskname2 = str(dictTasksDetails["Subtask"]).encode('utf-8').decode('utf-8').strip()
 
+# Function to prepare project sheet with spotlight data
+def prepareProjectAndTasksSheetsForSpotlight(project_inputFile, projectName_for_folder_path, accessToken, signedUrl):
+    millisecond = int(time.time() * 1000)
```
[the remark this answers is not in the document]
Comment (adithyadinesh0412) [reply Project-Service-implementation-Script/main.py:3437]: and test @MallanagoudaB

File: Project-Service-implementation-Script/main.py [hunk 4]
```diff
@@ -3430,9 +3431,240 @@ def prepareProjectAndTasksSheets(project_inputFile, projectName_for_folder_path,
...
+        if existing_data:
+            existing_header = existing_data[0]
+            existing_rows = existing_data[1:]
+        else:
+            existing_header = []
+            existing_rows = []
+    else:
+        # Initialize empty header and rows if file doesn't exist
+        existing_header = []
+        existing_rows = []
```
Comment (adithyadinesh0412) [inline Project-Service-implementation-Script/main.py:3460]: @MallanagoudaB As discussed. please optimise this

File: Project-Service-implementation-Script/main.py [hunk 5]
```diff
@@ -3430,9 +3431,240 @@ def prepareProjectAndTasksSheets(project_inputFile, projectName_for_folder_path,
...
+            dynamic_columns.append(key.strip())
+
+    # Combine base columns and dynamic columns
+    all_columns = new_columns + dynamic_columns
+
+    # Add new columns to header if they don't exist
+    for column in all_columns:
+        if column not in existing_header:
+            existing_header.append(column)
+            for row in existing_rows:
```
Comment (adithyadinesh0412) [inline Project-Service-implementation-Script/main.py:3479]: recheck this not convienced with the logic @MallanagoudaB

File: Project-Service-implementation-Script/main.py [hunk 6]
```diff
@@ -4196,6 +4430,149 @@ def solutionCreationAndMapping(projectName_for_folder_path, entityToUpload, list
...
+            detailsEnvSheet = wbproject.sheet_by_name(prosheet)
+            
+            # Read the header row to identify evidence columns
+            keysEnv = [
+                detailsEnvSheet.cell(1, col_index_env).value
+                for col_index_env in range(detailsEnvSheet.ncols)
+            ]
+            
+            # Create a folder to store evidence files
+            Logofilepath = os.path.join(projectName_for_folder_path, 'evidenceFile')
```
Comment (adithyadinesh0412) [inline Project-Service-implementation-Script/main.py:4451]: pls give meaningful name for variables @MallanagoudaB

File: Project-Service-implementation-Script/main.py [hunk 7]
```diff
@@ -4634,6 +5017,11 @@ def addProjectFunc(filePathAddProject, projectName_for_folder_path, millisAddObs
                                     downloadlogosign(filePathAddProject,projectName_for_folder_path)
                                     editsvg(accessToken,filePathAddProject,projectName_for_folder_path,baseTemplate_id)
                                     prepareProjectAndTasksSheets(addObservationSolution, projectName_for_folder_path,accessToken)
+                                    downloadEvidences(filePathAddProject,projectName_for_folder_path)
```
Comment (adithyadinesh0412) [inline Project-Service-implementation-Script/main.py:5020]: handle failure case @MallanagoudaB

**Review by VISHNUDAS-tunerlabs — APPROVED** [review]

**Pull request merged**